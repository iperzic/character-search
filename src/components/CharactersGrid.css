.CharactersGrid {
    display: grid;
    grid-gap: 20px;
    padding: 0 20px;
    margin-bottom: 15px;
}

@media (min-width: 1281px) {
    .CharactersGrid {
        grid-template-columns: repeat(5, minmax(200px, 1fr));
    }
}

@media (min-width: 768px) and (max-width: 1280px) {
    .CharactersGrid {
        grid-template-columns: repeat(2, minmax(200px, 1fr));
    }
}
