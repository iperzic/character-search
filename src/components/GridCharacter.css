.GridCharacter {
    position: relative;
    height: 220px;
    width: 100%;
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 4px 20px 0 rgba(0, 0, 0, 0.19);
}

.GridCharacter__image {
    height: 220px;
    width: 100%;
    object-fit: cover;
}

.GridCharacter__name {
    position: absolute;
    bottom: 0;
    width: calc(100% - 20px);
    padding: 10px;
    background-color: rgba(255, 255, 255, 0.4);
    margin-bottom: 0;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}

.GridCharacter__bookmark {
    position: absolute;
    top: 0;
    right: 0;
    padding: 5px;
    cursor: pointer;
    font-size: 1.2em;
    color: gold;
    text-shadow: -1px 0 black, 0 1px black, 1px 0 black, 0 -1px black;
}

.GridCharacter__bookmark:hover {
    opacity: 0.7;
}

.bookmark--not-selected:after {
    content: '☆';
}

.bookmark--selected:after {
    content: '★';
}
