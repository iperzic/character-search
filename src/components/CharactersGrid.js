import React from 'react';
import PropTypes from 'prop-types';

import './CharactersGrid.css';

import GridCharacter from './GridCharacter';

const CharactersGrid = ({ characters, onToggleBookmark }) => (
  <div className="CharactersGrid">
    {characters.map(c => (
      <GridCharacter
        key={c.id}
        name={c.name}
        image={c.image}
        isBookmark={c.isBookmark}
        onBookmarkClick={() => onToggleBookmark(c.id)}
      />
    ))}
  </div>
);

CharactersGrid.propTypes = {
  characters: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      image: PropTypes.string.isRequired,
      isBookmark: PropTypes.bool.isRequired,
    }).isRequired).isRequired,
  onToggleBookmark: PropTypes.func.isRequired,
};

export default CharactersGrid;
