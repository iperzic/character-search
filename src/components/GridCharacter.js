import React from 'react';
import PropTypes from 'prop-types';

import './GridCharacter.css';

const GridCharacter = ({ name, image, isBookmark, onBookmarkClick }) => (
  <div className="GridCharacter">
    <img className="GridCharacter__image" src={image} alt={name} />
    <h2 className="GridCharacter__name" title={name}>{name}</h2>
    <span
      className={isBookmark ?
        'GridCharacter__bookmark bookmark--selected' :
        'GridCharacter__bookmark bookmark--not-selected'}
      onClick={onBookmarkClick}
    />
  </div>
);

GridCharacter.propTypes = {
  name: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  isBookmark: PropTypes.bool.isRequired,
  onBookmarkClick: PropTypes.func.isRequired,
};

export default GridCharacter;
