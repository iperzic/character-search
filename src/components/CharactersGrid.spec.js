import React from 'react';
import { shallow } from 'enzyme';

import CharactersGrid from './CharactersGrid';
import GridCharacter from './GridCharacter';

describe('CharactersGrid', () => {
  let component;
  let toggleBookmarkSpy;
  let characters;

  beforeEach(() => {
    toggleBookmarkSpy = jest.fn();
    characters = [{
      id: 101,
      name: 'Fictional character 1',
      image: 'http://fictional-character-1-url',
      isBookmark: false,
    }, {
      id: 102,
      name: 'Fictional character 2',
      image: 'http://fictional-character-2-url',
      isBookmark: false,
    }, {
      id: 103,
      name: 'Fictional character 3',
      image: 'http://fictional-character-3-url',
      isBookmark: false,
    }];

    component = shallow(<CharactersGrid characters={characters} onToggleBookmark={toggleBookmarkSpy} />)
  });

  it('should display all provided characters', () => {
    expect(component.getElement()).toMatchSnapshot();
  });

  it('should call the onToggleBookmark prop when a character is clicked', () => {
    const index = 0;
    const gridCharacterComponent = component.find(GridCharacter).at(index);

    gridCharacterComponent.dive().find('.GridCharacter__bookmark').simulate('click');

    expect(toggleBookmarkSpy).toHaveBeenCalledWith(characters[index].id)
  });
});
