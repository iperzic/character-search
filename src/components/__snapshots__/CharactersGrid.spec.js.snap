// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharactersGrid should display all provided characters 1`] = `
<div
  className="CharactersGrid"
>
  <GridCharacter
    image="http://fictional-character-1-url"
    isBookmark={false}
    name="Fictional character 1"
    onBookmarkClick={[Function]}
  />
  <GridCharacter
    image="http://fictional-character-2-url"
    isBookmark={false}
    name="Fictional character 2"
    onBookmarkClick={[Function]}
  />
  <GridCharacter
    image="http://fictional-character-3-url"
    isBookmark={false}
    name="Fictional character 3"
    onBookmarkClick={[Function]}
  />
</div>
`;
